fixes #669: cleaning up code for entity worker

ktuite opened the pull request:
Closes #669 by cleaning up the entity creation code used by the worker that processes submission update events. 

- went back to relying on postgres to flag entity creation errors due to constraint violations (use a nested transaction to do this and properly catch the errors and then be able to continue logging)
- made the processing code use async/await to be easier to follow
- moved the audit logging code for entity creation and error events into the processing code, not the worker
- refactored the tests to be clearer and handle different kinds of failure/non-failure cases:
    * submission should not be processed
    * submission successfully makes entity
    * submission fails to make entity because of validation error
    * submission fails to make entity because of constraint violation error (duplicate uuid or something)

<!-- 
Thank you for contributing to ODK Central!

Before sending this PR, please read
https://github.com/getodk/central-backend/blob/master/CONTRIBUTING.md
-->

#### What has been done to verify that this works as intended?

Tests.

#### Why is this the best possible solution? Were any other approaches considered?

It's a lot better than it was! This change gets rid of a big weird hacky code block that was trying to carefully catch specific kind of errors.

#### How does this change affect users? Describe intentional changes to behavior and behavior that could have accidentally been affected by code changes. In other words, what are the regression risks?

n/a

#### Does this change require updates to the API documentation? If so, please update docs/api.md as part of this PR.

no

#### Before submitting this PR, please make sure you have:

- [x] run `make test-full` and confirmed all checks still pass OR confirm CircleCI build passes

ktuite commented on lib/model/query/entities.js:
```diff
@@ -62,55 +62,49 @@ select ins.*, def.id as "entityDefId", ds."acteeId" as "dsActeeId", ds."name" as
+    // and don't log it as an error.
+    if (existingEntity.isDefined())
+      return null;
+    const submissionDef = await Submissions.getDefById(submissionDefId).then((s) => s.get());
```
Possibly add test: What if the submission def doesn't exist?

ktuite commented on lib/model/query/entities.js:
```diff
@@ -62,55 +62,49 @@ select ins.*, def.id as "entityDefId", ds."acteeId" as "dsActeeId", ds."name" as
+    const entity = await Entities.createNew(partial, submissionDef);
+    return entity;
+  } catch (problem) {
+    if (problem.problemCode === 409.14)
```
Clarify with comment that these are all mainly entity validation errors

ktuite replied:
I've switched to catching the errors more generically so this comment isn't relevant anymore.

ktuite reviewed (commented):
Interactive review

Question: can we make a test to emulate DB failure to see worker try again?

matthew-white commented:
> Question: can we make a test to emulate DB failure to see worker try again?

Could you modify the `container` within the test? Something like:

```js
container.Entities.getDefBySubmissionDefId = () => reject(new Error('completely unhandled error'));
```

ktuite commented on lib/model/query/entities.js:
```diff
@@ -62,55 +62,49 @@ select ins.*, def.id as "entityDefId", ds."acteeId" as "dsActeeId", ds."name" as
+    if (problem.problemCode === 409.14)
+      throw problem;
+    else
+      throw Problem.user.entityViolation({ reason: 'Something went wrong validating or inserting' });
```
classify errors a bit more to distinguish between system errors and data errors
- uuid clash
- value of something is not in another table as expected
- some piece of data is missing

ktuite replied:
if we don't make this new problem, but we just pass whatever other error we get through, make sure that in the outer logging step, we strip out the stack trace if it is there.

ktuite reviewed (commented):
Interactive review

Question: can we make a test to emulate DB failure to see worker try again?

ktuite commented on lib/model/query/entities.js:
```diff
@@ -62,55 +62,49 @@ select ins.*, def.id as "entityDefId", ds."acteeId" as "dsActeeId", ds."name" as
+      container.Audits.log({ id: event.actorId }, 'entity.create.error', null,
+        { submissionId: event.details.submissionId,
+          submissionDefId: event.details.submissionDefId,
+          problem }));
```
does this log the stack trace?

ktuite replied:
it does not log the stack trace.

ktuite reviewed (commented):
interactive review with @matthew-white

ktuite commented:
I've updated the code to catch and log whatever kind of error the inner `processSubmissionEvent` might throw.  I've also changed the tests to reflect this.

Now you can see when a database uniqueness Problem comes through, or even an error that's not one of our Problems but something like a slonik error.

I haven't figured out an even deeper way of messing with the container to trigger an error. Deleting database tables or breaking the submission event all cause errors that get caught in a reasonable way.

I just wanted to write out another thought behind this error handling. In many other cases of workers doing things in Central, there's a clear reason to retry the task if there is a problem. With entities, if there is any sort of problem processing the submission, it seems okay to not retry, and to rely on the user to notice and re-approve the submission to get it to reprocess. At least for this first pass of entities while we discover what kinds of failure modes are possible.

matthew-white commented on lib/model/query/entities.js:
```diff
@@ -62,55 +61,52 @@ select ins.*, def.id as "entityDefId", ds."acteeId" as "dsActeeId", ds."name" as
+      container.Audits.log({ id: event.actorId }, 'entity.create.error', null,
+        { submissionId: event.details.submissionId,
+          submissionDefId: event.details.submissionDefId,
+          errorMessage: err.message,
```
It'd be nice to show this message in Frontend at some point in case the error isn't a Problem. I don't think we need to do that for v2022.3 though. Let me know if it'd be useful for me to file a Frontend issue about this!

ktuite replied:
Yes! we should be able to update frontend to check for `errorMessage` instead. It will have slightly more detail (include the name of the problem concatenated with the reason) than just the currently displayed Problem reason.
